core/application/bloodDonationWorkflow/BloodDonationService.ts

```diff
@@ -202,71 +196,87 @@ export class BloodDonationService {
 
   async routeDonorRequest(
     donorRoutingAttributes: DonorRoutingAttributes,
-    bloodDonationRepository: Repository<DonationDTO>,
-    stepFunctionModel: StepFunctionModel,
+    queueSource: string,
+    userProfile: UserDetailsDTO,
     donorSearchRepository: Repository<DonorSearchDTO>,
-    userRepository: Repository<UserDetailsDTO>
+    stepFunctionModel: StepFunctionModel
   ): Promise<string> {
     try {
       const { seekerId, requestPostId, createdAt } = donorRoutingAttributes
-      const bloodDonationItem = await bloodDonationRepository.getItem(
-        `${BLOOD_REQUEST_PK_PREFIX}#${seekerId}`,
-        `${BLOOD_REQUEST_PK_PREFIX}#${createdAt}#${requestPostId}`
-      )
-      if (bloodDonationItem === null) {
-        return 'Item not found.'
-      }
-      if (bloodDonationItem.status === DonationStatus.COMPLETED || bloodDonationItem.status === DonationStatus.EXPIRED) {
-        return 'You can\'t update the donation request'
-      }
 
       const donorSearchItem = await donorSearchRepository.getItem(
         `${DONOR_SEARCH_PK_PREFIX}#${seekerId}`,
         `${DONOR_SEARCH_PK_PREFIX}#${createdAt}#${requestPostId}`
       )
+
       if (donorSearchItem === null) {
-        await donorSearchRepository.create(bloodDonationItem)
+        await donorSearchRepository.create({
+          id: generateUniqueID(),
+          ...donorRoutingAttributes,
+          status: DonationStatus.PENDING,
+          retryCount: 0
+        })
       } else if (donorSearchItem.status === DonationStatus.COMPLETED) {
-        return 'Donor search is completed'
+        if (
+          queueSource === process.env.DONOR_SEARCH_QUEUE_ARN &&
+          donorRoutingAttributes.bloodQuantity > donorSearchItem.bloodQuantity
+        ) {
+          const updateData: Partial<DonorSearchDTO> = {
+            ...donorRoutingAttributes,
+            id: requestPostId,
+            status: DonationStatus.PENDING,
+            retryCount: 0
+          }
+          await donorSearchRepository.update(updateData)
+        } else {
+          return 'Donor search is completed'
```

Comment:
the promise is a string, this only returns a string when it hits this else block

Reply:
yes. and if it goes to if block it will continue going down.

the scenario is , if the donor search entry for this request has status `complete` it will check if it comes from the donor search query(which triggers because of create and update post) and it will check if the blood quantity is updated. if true it will change the status and retry count for donor search entry. and continue to the next steps